servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java

```diff
@@ -391,6 +484,16 @@ boolean addConnection(C connection) {
                     }
                     if (i == existing.length) {
                         break;
+                    } else if (existing.length == 1 && isExpired) {
+                        // We're closing the last connection, close the Host.
+                        // Closing the host will trigger the Host's onClose method,
+                        // which will remove the host from active hosts list.
+                        // If a race condition appears and a new connection was added in the meantime,
+                        // the CAS operation will allow for determining that and not closing the Host yet and removing
+                        // the connection, previously considered as the last one, from the array.
+                        if (connectionsUpdater.compareAndSet(this, existing, CLOSED_ARRAY)) {
```

Comment:
Actually, I was thinking about this race one more time. Let's draw an example:
```
thread-a: close the last connection for this expired host, we go inside `existing.length == 1 && isExpired` branch
sd-thread: mark this host as available
thread-b: select this host to open a new connection
thread-a: swaps the "connections" array with CLOSED_ARRAY, removes this host from the list of active hosts
thread-b: fails to add a new connection because the host is already closed, closes the new connection
```
I'm ok with the `thread-b` closing the newly created connection. But it doesn't look good if we remove the host from the list of active hosts that recently became available 🤔 
WDYT if instead of a boolean `isExpired` we will have an enum `state` variable with 3 possible values: ACTIVE, EXPIRED, CLOSING?
Your previous `if` check will change to:
```java
...
} else if (existing.length == 1) {
    if (stateUpdater.compareAndSet(this, EXPIRED, CLOSING)) {
        this.closeAsync().subscribe();
        break;
    }
    if (connectionsUpdater.compareAndSet(this, existing, EMPTY_ARRAY)) {
        break;
    }
} else {
...
```
It will help to avoid a race between `thread-a` and `sd-thread`. `sd-thread` will be able to see that the current host is closing and will add a new entry for this address. The old host entry will be removed by `thread-a`. The internal state won't have duplicates.

Reply:
I see your point, thank you for bringing this up. While addressing your concern about duplicates I tried to encapsulate Host's state a bit and that led me to come up with an idea for avoiding the state machine. Possibly the state machine would be helpful if the SD events were not serialised, but as I understand they currently are, which means we can use that assumption to our advantage. I came up with with a guarding mechanism in the form of a new method called `tryToMarkActive`. It uses CAS on the `connectionsUpdater` to determine CLOSING/CLOSED state and will instruct the caller to add a new host when that's the case. New commit coming right up, please have a look @idelpivnitskiy.

Reply:
> if the SD events were not serialised, but as I understand they currently are, which means we can use that assumption to our advantage

Sure, SD-events are emitted serially. The problem is that there is a race between thread that closes the connection and a thread that emits a new SD-event. You can discard `thread-b` from my example above to simplify it.

Doesn't look like current variant of `tryToMarkActive()` helps because it can be executed on `sd-thread` in between `existing.length == 1 && isExpired` (line 526) and `connectionsUpdater.compareAndSet(this, existing, CLOSED_ARRAY)` (line 533). I've added a test, see `closureOfLastConnectionDoesntRaceWithNewAvailableEvent()`. But because it's a concurrency test, it does not fail all the time. Use "repeat until failure" feature in Intellij IDEA. For me it fails within ~10k iterations.